### ISSUE TITLE
MissingMethodException when using ASP.NET Core in .NET framework

### ISSUE THREAD

**Issue by jskeet** [GoogleCloudPlatform/google-cloud-dotnet#1448]
- Create new ASP.NET Core 2.0 project targeting .NET 4.6.1
- Add NuGet reference to Google.Cloud.Diagnostics.AspNetCore 1.0.0 (see #1449 for details...)
- In Startup.cs:
  - Add a using directive: `using Google.Cloud.Diagnostics.AspNetCore;`
  - In Configure, add `app.UseGoogleTrace();`
  - In ConfigureServices, add `services.AddGoogleTrace(options => options.ProjectId = "your-project-id-here");`
- In HomeController.cs
  - Add a constructor: `public HomeController(TraceHeaderContext context) {}` and add the corresponding using directive (`using Google.Cloud.Diagnostics.Common;`)
- Run...

Exception:

```text
MissingMethodException: Method not found: 'System.Nullable`1<Boolean>
Google.Cloud.Diagnostics.Common.TraceDecisionPredicate.ShouldTrace(Microsoft.AspNetCore.Http.HttpRequest)'.
```

The problem is that the netstandard build of Google.Cloud.Diagnostics.Common.TraceDecisionPredicate uses Microsoft.AspNetCore.Http.HttpRequest where the net45 build uses System.Web.HttpRequest. When an ASP.NET Core application runs on .NET 4.6.1, those two collide.

I'm not at all sure what we're going to do about this...

**Comment by jskeet** [comment]
We have the same kind of issue in `IExceptionLogger`

**Comment by jskeet** [comment]
This affects ASP.NET Core 1.1 as well. (I suspect ASP.NET Core 1.0 would fail too, but I can't get that to work at all with .NET 4.6.1 at the moment.)

**Comment by ILMTitan** [comment]
I can confirm that ASP.NET Core 1.0 fails as well.

**Comment by iantalarico** [comment]
@jskeet I may be missing something here but don't we just need to ensure we check for NET46 [here](https://github.com/GoogleCloudPlatform/google-cloud-dotnet/blob/master/apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/TraceDecisionPredicate.cs#L18)?

**Comment by jskeet** [comment]
@iantalarico: No, because that isn't checked at execution time - it's checked at compile time.

At execution time in the problematic scenario, we have:

- Your app, targeting net461 or whatever
- Google.Cloud.Diagnostics.AspNetCore targeting netstandard15, which has been compiled against Google.Cloud.Diagnostics.Common targeting netstandard15, so it thinks TraceDecisionPredicate uses Microsoft.AspNetCore.Http.HttpRequest
- Google.Cloud.Diagnostics.Common targeting net45 (because that's closest to the deployment platform) - where TraceDecisionPredicate uses System.Web.HttpRequest

Basically, we shouldn't have two "flavours" of a library in the same package which expose the same member but with different signatures.

**Comment by iantalarico** [comment]
Ah okay, so essentially we need to pull out all of this code into the respective libraries?

**Comment by jskeet** [comment]
Or maybe make it a generic type, yes. Something like that.
